sdk/nodejs/s3/routingRules.ts

```diff
@@ -0,0 +1,182 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * An array of `RoutingRule` objects.
+ *
+ * For more details, please refer to the AWS documentation online:
+ * https://docs.aws.amazon.com/AmazonS3/latest/dev/how-to-page-redirect.html#advanced-conditional-redirects
+ */
+export type RoutingRules = RoutingRule[];
+
+/**
+ * A rule that identifies a condition and the redirect that is applied when the condition is met.
+ * If a condition is not included, the rule is applied to all requests.
+ */
+export interface RoutingRule {
+    /**
+     * A condition that must be met for the specified redirect to be applied.
+     * If not included, the rule is applied to all requests.
+     */
+    Condition?: Condition;
+
+    /**
+     * Provides instructions for redirecting the request. You can redirect requests to another host or another page, or you can specify another protocol to use.
+     */
+    Redirect: Redirect;
+}
+
+// Multiple types are used for `Condition` to enforce at least one property being set.
+
+/**
+ * A condition that must be met for the specified redirect to be applied.
+ */
+export type Condition = HttpErrorCodeReturnedEqualsCondition | KeyPrefixEqualsCondition;
+
+export interface HttpErrorCodeReturnedEqualsCondition {
+    /**
+     * The HTTP error code that must match for the redirect to apply. If an error occurs, and if the error code meets this value, then the specified redirect applies.
+     */
+    HttpErrorCodeReturnedEquals: string;
+
+    /**
+     * The prefix of the object key name from which requests are redirected.
+     *
+     * If specified, both `HttpErrorCodeReturnedEquals` and `KeyPrefixEquals` must be true for the condition to be met.
+     */
+    KeyPrefixEquals?: string;
+}
+
+export interface KeyPrefixEqualsCondition {
+    /**
+     * The HTTP error code that must match for the redirect to apply. If an error occurs, and if the error code meets this value, then the specified redirect applies.
+     *
+     * If specified, both `HttpErrorCodeReturnedEquals` and `KeyPrefixEquals` must be true for the condition to be met.
+     */
+    HttpErrorCodeReturnedEquals?: string;
+
+    /**
+     * The prefix of the object key name from which requests are redirected.
+     */
+    KeyPrefixEquals: string;
+}
+
+// Multiple types are used for `Redirect` to enforce at least one property being set.
```

Comment:
this feels contorted to me.  i woudl just make optional and have a runtime check TBH.

Reply:
Would this be a runtime check we implement (if so how?), or just leave it up to AWS to do any checking?

Reply:
Yes.  I twould be something like `if (a.x === undefined && a.y === undefined && a.z === undefined) { throw new ResourceError("At least one of blah blah blah must be set"); }`

Reply:
Yeah, but where would such a check go? Is there a place on the TS/JS side to put it?

I suppose I could write it on the Go side... instead of defining the field as:

```go
Fields: map[string]*tfbridge.SchemaInfo{
        "routing_rules": {
            Name:      "routingRules",
            Type:      "string",
            AltTypes:  []tokens.Type{awsType(s3Mod+"/routingRules", "RoutingRules")},
            Transform: tfbridge.TransformJSONDocument,
        },
    },
```

I could write a custom `Transform` that wraps the call to `TransformJSONDocument` and adds custom validation, e.g. something like:

```go
Transform: func(val resource.PropertyValue) (resource.PropertyValue, error) {
	result, err := tfbridge.TransformJSONDocument(val)
	if err != nil {
		return resource.PropertyValue{}, err
	}

	// validate val here

	// if ok
	return result, nil
},
```

But that seems kind of cumbersome.

Reply:
@CyrusNajmabadi?

Reply:
Answered offline.